Add logging to WTF::SuspendableWorkQueue

szewai opened the pull request:
#### 4b3bce8b1e4f520b429c5ad943f3c62be3536e66
<pre>
Add logging to WTF::SuspendableWorkQueue
<a href="https://bugs.webkit.org/show_bug.cgi?id=250809">https://bugs.webkit.org/show_bug.cgi?id=250809</a>
rdar://problem/104403363

Reviewed by Chris Dumez.

To help debug storage queue suspension issues like rdar://103873458.

* Source/WTF/wtf/Logging.cpp:
* Source/WTF/wtf/Logging.h:
* Source/WTF/wtf/SuspendableWorkQueue.cpp:
(WTF::SuspendableWorkQueue::create):
(WTF::SuspendableWorkQueue::SuspendableWorkQueue):
(WTF::SuspendableWorkQueue::stateString):
(WTF::SuspendableWorkQueue::suspend):
(WTF::SuspendableWorkQueue::resume):
(WTF::SuspendableWorkQueue::suspendIfNeeded):
* Source/WTF/wtf/SuspendableWorkQueue.h:
* Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp:
(WebKit::NetworkStorageManager::NetworkStorageManager):
(WebKit::NetworkStorageManager::suspend):
(WebKit::NetworkStorageManager::resume):

Canonical link: <a href="https://commits.webkit.org/259130@main">https://commits.webkit.org/259130@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/3a32de92665f48c5693cdffd175441344c548e7e

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/103981 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/13088 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/36917 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/113202 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/173508 "Built successfully and passed tests") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/14128 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/3993 "Built successfully") | [❌ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/96223 "Failed to checkout and rebase branch from PR 8807") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/112293 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/109753 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/10900 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/93957 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/36/builds/96223 "Failed to checkout and rebase branch from PR 8807") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/92739 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/25568 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/36/builds/96223 "Failed to checkout and rebase branch from PR 8807") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/94064 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/6452 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/26949 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/90562 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/83/builds/4225 "Built successfully and passed tests") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/6605 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/3491 "Found 1 new test failure: fast/images/avif-image-document.html (failure)") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/29883 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/12598 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/46479 "Passed tests") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/99170 "Built successfully") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/6289 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/8379 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/24955 "Passed tests") | 
| | | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WTF/wtf/SuspendableWorkQueue.cpp:
```diff
@@ -26,33 +26,53 @@
     ASSERT(isMainThread());
 }
 
+void SuspendableWorkQueue::log(ASCIILiteral message)
```
Please use `RELEASE_LOG_IF(m_shouldLog)` instead of this whole function.
You can also create your own new RELEASE_LOG_SWQ() macro in this file which wraps the  `RELEASE_LOG_IF(m_shouldLog)` call to simplify call sites.

Release logging should be cheap and thus should avoid function calls.

cdumez commented on Source/WTF/wtf/SuspendableWorkQueue.cpp:
```diff
@@ -55,16 +74,15 @@ void SuspendableWorkQueue::suspend(Function<void()>&& suspendFunction, Completio
     ASSERT(isMainThread());
     Locker suspensionLocker { m_suspensionLock };
 
+    RELEASE_LOG_IF(m_shouldLog, SuspendableWorkQueue, "%p - SuspendableWorkQueue::resume current state %s", this, stateString(m_state));
```
This needs to use:
`current state %" PUBLIC_LOG_STRING`

Otherwise, it won't show in non-internal sysdiagnoses.

cdumez commented on Source/WTF/wtf/SuspendableWorkQueue.cpp:
```diff
@@ -26,24 +26,43 @@
     ASSERT(isMainThread());
 }
 
+const char* SuspendableWorkQueue::stateString(State state)
```
I'd recommend to define this only in the cpp (since you don't need outside this class and mark as inline:
```
static inline const char* SuspendableWorkQueue::stateString(State state)
{
...
}
```

Just to be extra sure we don't generate a function call.

cdumez commented on Source/WTF/wtf/SuspendableWorkQueue.cpp:
```diff
@@ -55,16 +74,15 @@ void SuspendableWorkQueue::suspend(Function<void()>&& suspendFunction, Completio
-    WorkQueue::dispatch([this] {
-        suspendIfNeeded();
-    });
+    dispatch([]() { });
```
What's this change? This doesn't look logging related.

szewai replied:
Will revert.

cdumez commented on Source/WTF/wtf/SuspendableWorkQueue.cpp:
```diff
@@ -26,24 +26,43 @@
     ASSERT(isMainThread());
     Locker suspensionLocker { m_suspensionLock };
 
+    RELEASE_LOG_IF(m_shouldLog, SuspendableWorkQueue, "%p - SuspendableWorkQueue::suspend current state %s", this, stateString(m_state));
```
Should use PUBLIC_LOG_STRING like I suggested below.

cdumez commented on Source/WTF/wtf/SuspendableWorkQueue.cpp:
```diff
@@ -120,18 +138,26 @@ void SuspendableWorkQueue::invokeAllSuspensionCompletionHandlers()
+    if (m_state != State::WillSuspend) {
+        if (m_state == State::Suspended) {
+            // If state is suspended, the queue should not be running.
+            RELEASE_LOG_ERROR_IF(m_shouldLog, SuspendableWorkQueue, "%p - SuspendableWorkQueue::suspendIfNeeded current state Suspended", this);
```
Could use `RELEASE_LOG_ERROR_IF(m_shouldLog && m_state == State::Suspended, ` to make the code a bit more concise.

cdumez commented on Source/WTF/wtf/SuspendableWorkQueue.h:
```diff
@@ -36,7 +36,9 @@ namespace WTF {
 public:
     using QOS = WorkQueue::QOS;
-    WTF_EXPORT_PRIVATE static Ref<SuspendableWorkQueue> create(const char* name, QOS = QOS::Default);
+    enum class State : uint8_t { Running, WillSuspend, Suspended };
```
Can we put it back where it was? Not sure why this needs to be public now.

szewai replied:
Because `static inline const char* stateString(SuspendableWorkQueue::State state)` in SuspendableWorkQueue.cpp?
Or do you mean I should change `stateString()` back to a class member?

cdumez replied:
Oh I see. Yes, given this I'd prefer if we put the static function declaration in the header as private (so State can stay private). Just mark it as inline in the cpp.

cdumez reviewed (approved):
LGTM with nit.

szewai committed:
Add logging to WTF::SuspendableWorkQueue
https://bugs.webkit.org/show_bug.cgi?id=250809
rdar://problem/104403363

Reviewed by Chris Dumez.

To help debug storage queue suspension issues like rdar://103873458.

* Source/WTF/wtf/Logging.cpp:
* Source/WTF/wtf/Logging.h:
* Source/WTF/wtf/SuspendableWorkQueue.cpp:
(WTF::SuspendableWorkQueue::create):
(WTF::SuspendableWorkQueue::SuspendableWorkQueue):
(WTF::SuspendableWorkQueue::stateString):
(WTF::SuspendableWorkQueue::suspend):
(WTF::SuspendableWorkQueue::resume):
(WTF::SuspendableWorkQueue::suspendIfNeeded):
* Source/WTF/wtf/SuspendableWorkQueue.h:
* Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp:
(WebKit::NetworkStorageManager::NetworkStorageManager):
(WebKit::NetworkStorageManager::suspend):
(WebKit::NetworkStorageManager::resume):

Canonical link: https://commits.webkit.org/259130@main
